### PR TITLE
fix error in PaginatorHelper with scopes

### PR DESCRIPTION
**Pull request by inmun**
This pull request references issue #15566

### CODE REVIEW

File: tests/TestCase/View/Helper/PaginatorHelperTest.php [hunk 1]
```diff
@@ -3495,4 +3495,32 @@ protected function setPagingParams($params, bool $merge = true)
...
+     * Test that generated URLs work without sort defined within the request
+     *
+     * @return void
+     */
+    public function testPaginationWithPassedParams()
+    {
+        $request = new ServerRequest([
+            'url' => '/articles/index/whatever/3',
+            'params' => [
+                'plugin' => null, 'controller' => 'articles', 'action' => 'index', 'pass' => ['whatever', '3']
```
Comment (dereuromark) [inline tests/TestCase/View/Helper/PaginatorHelperTest.php:3509]:
```suggestion
                'plugin' => null, 'controller' => 'Articles', 'action' => 'index', 'pass' => ['whatever', '3']
```

Comment (inmun) [reply tests/TestCase/View/Helper/PaginatorHelperTest.php]: If controller param is capitalized the url result changes to: Articles/index...

File: tests/TestCase/View/Helper/PaginatorHelperTest.php [hunk 2]
```diff
@@ -3495,4 +3495,35 @@ protected function setPagingParams($params, bool $merge = true)
...
+     *
+     * @return void
+     */
+    public function testPaginationWithPassedParams()
+    {
+        $request = new ServerRequest([
+            'url' => '/articles/index/whatever/3',
+            'params' => [
+                'plugin' => null,
+                'controller' => 'articles',
```
Comment (dereuromark) [inline tests/TestCase/View/Helper/PaginatorHelperTest.php:3510]: Can this be Articles as well? Or does the test then fail?

Comment (dereuromark) [reply tests/TestCase/View/Helper/PaginatorHelperTest.php:3510]: Refs https://github.com/cakephp/cakephp/pull/15568

Comment (inmun) [reply tests/TestCase/View/Helper/PaginatorHelperTest.php:3510]: The test fails. The url result changes to capitalize. A similar test with the same problem [here].(https://github.com/cakephp/cakephp/blob/master/tests/TestCase/View/Helper/PaginatorHelperTest.php#L817)
I don´t know what's happening.

Comment (dereuromark) [reply tests/TestCase/View/Helper/PaginatorHelperTest.php:3510]: OK; we can tackle this in a follow up PR, this one is fine then.

File: src/View/Helper/PaginatorHelper.php [hunk 3]
```diff
@@ -567,20 +567,18 @@ public function generateUrlParams(array $options = [], ?string $model = null, ar
...
-            if (isset($currentParams['?'][$scope]) && is_array($currentParams['?'][$scope])) {
-                $options += $currentParams['?'][$scope];
+            if (isset($baseUrl['?'][$scope]) && is_array($baseUrl['?'][$scope])) {
+                $options += $baseUrl['?'][$scope];
+                unset($baseUrl['?'][$scope]);
             }
             $options = [$scope => $options];
         }
 
         if (!empty($this->_config['options']['url'])) {
```
Comment (markstory) [inline src/View/Helper/PaginatorHelper.php:580]: Should this line also be `!empty($baseUrl)`?

Comment (othercorey) [reply src/View/Helper/PaginatorHelper.php:580]: @inmun Can you follow up so we can merge or close this?

File: tests/TestCase/View/Helper/PaginatorHelperTest.php [hunk 4]
```diff
@@ -3495,4 +3495,35 @@ protected function setPagingParams($params, bool $merge = true)
...
+
+        $this->View->setRequest($request->withAttribute('paging', [
+            'Article' => [
+                'scope' => 'article',
+            ],
+        ]));
+        $this->Paginator = new PaginatorHelper($this->View);
+        $result = $this->Paginator->generateUrl(['page' => 2]);
+        $expected = '/articles/index/whatever/3?article%5Bpage%5D=2';
+        static::assertSame($expected, $result);
```
Comment (dereuromark) [inline tests/TestCase/View/Helper/PaginatorHelperTest.php:3527]:
```suggestion
        $this->assertSame($expected, $result);
```
is now this should be used

Comment (inmun) [reply tests/TestCase/View/Helper/PaginatorHelperTest.php:3527]: @dereuromark I changed it. There is no problem in this context, but it is not good practice to call a static method dynamically [https://github.com/sebastianbergmann/phpunit/blob/9.5/src/Framework/Assert.php#L1327](https://github.com/sebastianbergmann/phpunit/blob/9.5/src/Framework/Assert.php#L1327)

Comment (dereuromark) [reply tests/TestCase/View/Helper/PaginatorHelperTest.php:3527]: Well, it is PHPUnit's official documentation and standpoint to use them non-statically :)

Comment (inmun) [reply tests/TestCase/View/Helper/PaginatorHelperTest.php:3527]: [https://phpunit.readthedocs.io/en/9.5/assertions.html#appendixes-assertions](https://phpunit.readthedocs.io/en/9.5/assertions.html#appendixes-assertions)
the both uses are right :)

Comment (dereuromark) [reply tests/TestCase/View/Helper/PaginatorHelperTest.php:3527]: Not if you follow the rest of their docs :)

Comment (inmun) [reply tests/TestCase/View/Helper/PaginatorHelperTest.php:3527]: Totally agree with you

**Review by dereuromark — APPROVED** [review]

**Comment by markstory** [comment]
Thank you 👍

**Pull request merged**